pkg/podclient/baseclient.go

```diff
@@ -0,0 +1,69 @@
+package podclient
+
+import (
+	"bufio"
+	"context"
+	"fmt"
+
+	"github.com/iver-wharf/wharf-core/pkg/logger"
+	v1 "k8s.io/api/core/v1"
+	metav1 "k8s.io/apimachinery/pkg/apis/meta/v1"
+	"k8s.io/apimachinery/pkg/watch"
+	"k8s.io/client-go/kubernetes"
+	corev1 "k8s.io/client-go/kubernetes/typed/core/v1"
+)
+
+// BaseClient is a basic implementation of the Client interface.
+type BaseClient struct {
+	Namespace string
+	// restConfig *rest.Config
+	Clientset *kubernetes.Clientset
+	Pods      corev1.PodInterface
+	// events     corev1.EventInterface
+}
```

Comment:
OK it took some reading to understand. You're basically doing OOP styled inheritence here with base classes with this `BaseClient` type. This usage is highly incompatible with unit tests as you're embedding this type in the dependent types. With this, we cannot only unit test `provisioner` without also needing to test `podclient` at the same time.

What confuses me is that you're also abstracting it with an interface, `podclient.Client`, suggesting that it's OK to be implemented elsewhere such as to mock it in unit tests, but you never leave that spot open as you're other code is depending on the concrete `podclient.BaseClient` type.

However, the amount of code is so small right now. You're trying to de-duplicate maybe 20-40 lines of code between the `pkg/worker` and `pkg/provisioner` packages before even seeing what to de-duplicate.

I suggest to instead defer adding this `pkg/podclient`. For example by:

1. Skip this `pkg/podclient` package.
2. Implement `pkg/provisioner` having code duplicated with `pkg/worker`. This is your rough draft.
3. Get your implementation to comply with the issue description.
4. Clean up the code in `pkg/provisioner`.
5. Compare `pkg/provisioner` and `pkg/worker` to spot shared code, and only then start abstracting out the shared code, perhaps into a reinstated `pkg/podclient` package.

I'm not sold on this `pkg/podclient` even existing, so I think the above workflow would give us a better understanding of what this `pkg/podclient` package could be used for.

Reply:
Yeah, I see what you mean with the dependence here. Didn't think about that.

Going with the 1-5 you wrote there. 👍🏻